### PR TITLE
RHOAI-9817 TrustyAI data drift

### PR DESCRIPTION
**Pull request by aduquett**
Adding upstream docs for monitoring data drift with TrustyAI and more reorganization.

### CODE REVIEW

File: modules/deleting-a-bias-metric-using-dashboard.adoc [hunk 1]
```diff
@@ -4,7 +4,7 @@
 = Deleting a bias metric by using the dashboard
 
 [role='_abstract']
-The following procedure describes how to use the {productname-short} dashboard to delete a bias metric for a model.
+You can usegit  the {productname-short} dashboard to delete a bias metric for a model.
```
Comment (aduquett) [inline modules/deleting-a-bias-metric-using-dashboard.adoc:7]: You can use the

File: assemblies/configuring-trustyai.adoc [hunk 2]
```diff
@@ -2,16 +2,17 @@
...
 
-[id="setting-up-model-monitoring_{context}"]
-= Setting up model monitoring
+[id="configuring-trustyai_{context}"]
+= Configuring TrustyAI
 
 [role='_abstract']
 
-To set up model monitoring for data scientists to use in {productname-short}, a cluster administrator does the following tasks:
+To configure model monitoring with TrustyAI for data scientists to use in {productname-short}, a cluster administrator does the following tasks:
```
Comment (chtyler) [inline assemblies/configuring-trustyai.adoc:10]: If this module is aimed at cluster administrators, could we not use "you" instead of "cluster administrator"?

..."you must complete the following tasks".

File: assemblies/setting-up-trustyai-for-your-project.adoc [hunk 3]
```diff
@@ -0,0 +1,31 @@
+:_module-type: ASSEMBLY
+
+ifdef::context[:parent-context: {context}]
+
+[id="setting-up-trustyai-for-your-project_{context}"]
+= Setting up TrustyAI for your project
+
+[role='_abstract']
+
+To set up model monitoring with TrustyAI for a data science project, a data scientist does the following tasks:
```
Comment (chtyler) [inline assemblies/setting-up-trustyai-for-your-project.adoc:10]: Same kind of comment as before, if this is aimed at data scientists, should we not use "you"?

File: assemblies/setting-up-trustyai-for-your-project.adoc [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+* Label your data fields (optional)
+
+After setting up, a data scientist can create and view bias and data drift metrics for deployed models. 
+
+[NOTE]
+====
+Model monitoring with TrustyAI is only available on the ModelMesh-based _multi-model serving platform_. Model monitoring with TrustyAI is unavailable on the KServe-based _single-model serving platform_.
+====
+
+include::modules/authenticating-trustyai-service.adoc[leveloffset=+1]
```
Comment (chtyler) [inline assemblies/setting-up-trustyai-for-your-project.adoc:23]: filename does not contain the word "the"

File: modules/authenticating-trustyai-service.adoc [hunk 5]
```diff
@@ -4,7 +4,7 @@
 = Authenticating the TrustyAI service
 
 [role='_abstract']
-You can access TrustyAI service external endpoints only if you provide OAuth proxy (oauth-proxy) authentication. You must obtain a user token, or a token from a service account with sufficient privileges, and then pass the token to the TrustyAI service when using `curl` commands.
+To access TrustyAI service external endpoints, you need to provide OAuth proxy (oauth-proxy) authentication. You must obtain a user token, or a token from a service account with sufficient privileges, and then pass the token to the TrustyAI service when using `curl` commands.
```
Comment (chtyler) [inline modules/authenticating-trustyai-service.adoc:7]: "need to " could be just "must"

File: modules/authenticating-trustyai-service.adoc [hunk 6]
```diff
@@ -13,20 +13,25 @@ endif::[]
 ifdef::cloud-service[]
 * You installed the OpenShift command line interface (`oc`) as described in link:https://docs.openshift.com/dedicated/cli_reference/openshift_cli/getting-started-cli.html[Getting started with the CLI] (OpenShift Dedicated) or link:https://docs.openshift.com/rosa/cli_reference/openshift_cli/getting-started-cli.html[Getting started with the CLI] (Red Hat OpenShift Service on AWS)
 endif::[]
-* Your OpenShift cluster administrator has installed {productname-short}, added you as a user to the OpenShift cluster, and installed the TrustyAI service for the data science project where the models are deployed.
+* Your OpenShift cluster administrator added you as a user to the {openshift-platform} cluster and has installed the TrustyAI service for the data science project where the models are deployed.
```
Comment (chtyler) [inline modules/authenticating-trustyai-service.adoc:16]: "for the data science project where the models are deployed"

could also be

"for the data science project that contains the deployed models"

File: modules/creating-a-bias-metric-using-dashboard.adoc [hunk 7]
```diff
@@ -16,16 +16,9 @@ endif::[]
 
 * You are familiar with the specific data set schema and understand the names and meanings of the inputs and outputs.
 
-* Your OpenShift cluster administrator has configured monitoring for the model serving platform, enabled the TrustyAI component in the {productname-long} {install-package}, and installed the TrustyAI service for the data science project where the models are deployed.
+* Your OpenShift cluster administrator added you as a user to the {openshift-platform} cluster and has installed the TrustyAI service for the data science project where the models are deployed.
```
Comment (chtyler) [inline modules/creating-a-bias-metric-using-dashboard.adoc:19]: See previous comment about potentially changing the wording for this.

File: modules/creating-a-drift-metric-using-cli.adoc [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+[id='creating-a-drift-metric-using-cli_{context}']
+= Creating a data drift metric by using the CLI
+
+[role='_abstract']
+You can use the OpenShift command-line interface (CLI) to create a data drift metric for a model.
+
+.Prerequisites
+
+* You are familiar with the specific data set schema and understand the names and meanings of the inputs and outputs.
```
Comment (chtyler) [inline modules/creating-a-drift-metric-using-cli.adoc:11]: The following could be shorted:

"...understand the names and meanings of the inputs and outputs."

"...and understand the relevant inputs and outputs."

File: modules/installing-trustyai-service.adoc [hunk 9]
```diff
@@ -1,11 +1,11 @@
 :_module-type: CONCEPT
 
 [id='installing-trustyai-service_{context}']
-= Installing the TrustyAI service for a data science project
+= Installing the TrustyAI service for a project
 
 [role='_abstract']
 
-Install TrustyAI into your project and provide access to its features for all models deployed within that project so that data scientists can monitor models. An instance of the TrustyAI service is required for each data science project, or namespace, that contains models that your team wants to monitor. 
+Install the TrustyAI service on a data science project to provide access to its features for all models deployed within that project. An instance of the TrustyAI service is required for each data science project, or namespace, that contains models that your team wants to monitor.
```
Comment (chtyler) [inline modules/installing-trustyai-service.adoc:8]: "your team" or  just "you"?

File: modules/creating-a-drift-metric-using-cli.adoc [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+* Your OpenShift cluster administrator added you as a user to the {openshift-platform} cluster and has installed the TrustyAI service for the data science project where the models are deployed.
+
+* You set up TrustyAI for your data science project, as described in xref:setting-up-trustyai-for-your-project_monitor[Setting up TrustyAI for your project].
+
+.Procedure
+
+. In a terminal window, log in to the OpenShift cluster where {productname-short} is deployed.
++
+----
+oc login
```
Comment (chtyler) [inline modules/creating-a-drift-metric-using-cli.adoc:22]: Is this login command sufficient, or do they need to log in with the token again? 

$ oc login --token=__<token>__ --server=__<openshift_cluster_url>__

File: modules/labeling-data-fields.adoc [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+.. In the upper-right corner of the OpenShift web console, click your user name and select *Copy login command*. 
+.. After you have logged in, click *Display token*.
+.. Copy the *Log in with this token* command and paste it in the OpenShift command-line interface (CLI).
++
+[source,subs="+quotes"]
+----
+$ oc login --token=__<token>__ --server=__<openshift_cluster_url>__
+----
+
+. In the OpenShift CLI, find the route to the TrustyAI service:
```
Comment (chtyler) [inline modules/labeling-data-fields.adoc:28]: Is "find" the most suitable word here? How about "get" or "obtain"?

File: modules/labeling-data-fields.adoc [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+. To examine TrustyAI's model metadata, query the `/info` endpoint:
++
+[source]
+----
+curl -H "Authorization: Bearer ${TOKEN}" $TRUSTY_ROUTE/info | jq ".[0].data"
+----
++
+This outputs a JSON file containing the following information for each model:
+
+* The names, data types, and positions of fields in the input and output.
```
Comment (chtyler) [inline modules/labeling-data-fields.adoc:44]: could reword this to "of input fields and output fields".

File: modules/labeling-data-fields.adoc [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+* The observed values that these fields take.
```
Comment (chtyler) [inline modules/labeling-data-fields.adoc:45]: Not sure if this could be reworded to:

"The observed field values."

File: modules/labeling-data-fields.adoc [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+      \"predict-0\": \"Prediction 0\"
+    }
+  }"
+----
++
+For another example, see https://github.com/trustyai-explainability/odh-trustyai-demos/blob/main/2-BiasMonitoring/scripts/apply_name_mapping.sh.
+
+.Verification
+
+You should see a "Feature and output name mapping successfully applied" message.
```
Comment (chtyler) [inline modules/labeling-data-fields.adoc:75]: Avoid using the word "should". 

"You can see  a "Feature and output name mapping successfully applied" message.

or 

a "Feature and output name mapping successfully applied" message is displayed.

File: modules/overview-of-model-monitoring.adoc [hunk 15]
```diff
@@ -26,13 +26,11 @@ To ensure that machine-learning models are transparent, fair, and reliable, data
...
 
-ifdef::upstream[]
 Bias::
-Check for unfair patterns or biases in data and model predictions to ensure your model's decisions are fair and unbiased.
-endif::[]
+Check for unfair patterns or biases in data and model predictions to ensure your model's decisions are unbiased.
 
 Data drift::
-Detect changes in data over time by comparing the latest real-world data to the original training data. Comparing the data identifies shifts or deviations that could impact model performance, ensuring the model remains accurate and reliable.
+Detect changes in input data distributions over time by comparing the latest real-world data to the original training data. Comparing the data identifies shifts or deviations that could impact model performance, ensuring the model remains accurate and reliable.
```
Comment (chtyler) [inline modules/overview-of-model-monitoring.adoc:33]: ".., ensuring that the model remains accurate and reliable."

File: modules/sending-training-data-to-trustyai.adoc [hunk 16]
```diff
@@ -0,0 +1,61 @@
+:_module-type: PROCEDURE
+
+[id="sending-training-data-to-trustyai{context}"]
+= Sending training data to TrustyAI
+
+[role='_abstract']
+To use TrustyAI for bias monitoring or data drift detection, you need to send training data for your model to TrustyAI.
```
Comment (chtyler) [inline modules/sending-training-data-to-trustyai.adoc:7]: "must" instead of  "need to".

File: modules/sending-training-data-to-trustyai.adoc [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+. Send data to this endpoint. For more information, see the link:https://kserve.github.io/website/0.8/modelserving/inference_api/#server-metadata-response-json-object[KServe v2 Inference Protocol documentation].
+
+.Verification
+Follow these steps to view cluster metrics and verify that TrustyAI is receiving data. 
+
+. Log in to the {openshift-platform} web console.
+. Switch to the *Developer* perspective.
+. In the left menu, click *Observe*.
+. On the *Metrics* page, click the *Select query* list and select *Custom query*.
```
Comment (chtyler) [inline modules/sending-training-data-to-trustyai.adoc:39]: and then *Custom query*.

File: modules/viewing-drift-metrics.adoc [hunk 18]
```diff
@@ -0,0 +1,30 @@
+:_module-type: PROCEDURE
+
+[id="viewing-drift-metrics_{context}"]
+= Viewing data drift metrics for a model
```
Comment (chtyler) [inline modules/viewing-drift-metrics.adoc:4]: Should the title be file name and ID be changed to match the title here (or vice versa)?

**Review by chtyler — CHANGES_REQUESTED** [review]
Looks really good, just a few minor changes requested.

File: assemblies/configuring-trustyai.adoc (hunk 2 above)
Comment (aduquett) [reply assemblies/configuring-trustyai.adoc:10]: I agree this sounds strange, but I wanted to make it clear this section was for admins until we separate the books.

File: assemblies/setting-up-trustyai-for-your-project.adoc (hunk 3 above)
Comment (aduquett) [reply assemblies/setting-up-trustyai-for-your-project.adoc:10]: I agree this sounds strange, but I wanted to make it obvious this section was for non-admins until we separate the books.

File: modules/supported-drift-metrics.adoc [hunk 19]
```diff
@@ -0,0 +1,38 @@
+:_module-type: REFERENCE
+
+[id="supported-drift-metrics_{context}"]
+= Supported drift metrics
+
+{productname-long} supports the following data drift metrics.
```
Comment (chtyler) [inline modules/supported-drift-metrics.adoc:6]: Could add a colon after metrics instead of a full stop.

File: modules/supported-drift-metrics.adoc [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+
+The FourierMMD metric provides the probability that the data values in a test data set have drifted from the training data set distribution, assuming that the computed Maximum Mean Discrepancy (MMD) values are normally distributed. This metric works by comparing the empirical distributions of the data sets by using an MMD measure in the Fourier domain. 
++
+FourierMMD is useful for detecting subtle shifts in data distributions that might be overlooked by simpler statistical measures.
++
+This metric returns the probability that the distribution seen in the "real world" data has drifted from the reference data. The closer the value is to 1, the more likely there is to be significant drift.
+
+KSTest::
+
+The KSTest metric calculates two Kolmogorov-Smirnov tests for each column to determine whether the data sets are drawn from the same distributions. This metric works by measuring the maximum distance between the empirical cumulative distribution functions (CDFs) of the data sets, without assuming any specific underlying distribution.
```
Comment (chtyler) [inline modules/supported-drift-metrics.adoc:26]: You could omit "works by" from: 

 This metric **works by** measuring the maximum distance between the empirical cumulative distribution functions (CDFs) of the data sets, without assuming any specific underlying distribution.
 
 Instead use "This metric measures..."

Comment (chtyler) [inline modules/supported-drift-metrics.adoc:26]: Could we change the term "drawn from" to "are derived" ?

File: modules/supported-drift-metrics.adoc [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
++
+KSTest is useful for detecting changes in distribution shape, location, and scale.
++
+This metric returns the probability that the distribution seen in the "real world" data comes from the same distribution as the reference data. The closer the value is to 0, the more likely there is to be significant drift.
```
Comment (chtyler) [inline modules/supported-drift-metrics.adoc:30]: Similar thing here: 

"..comes from.." could be changed to " ..is derived from.."

File: modules/supported-drift-metrics.adoc [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+ApproxKSTest::
+
+The ApproxKSTest metric performs an approximate Kolmogorov-Smirnov test, ensuring that the maximum error is `6*epsilon` compared to an exact KSTest. 
++
+ApproxKSTest is useful for detecting changes in distributions for large data sets where performing an exact KSTest might be computationally expensive.
++
+This metric returns the probability that the distribution seen in the "real world" data comes from the same distribution as the reference data. The closer the value is to 0, the more likely there is to be significant drift.
```
Comment (chtyler) [inline modules/supported-drift-metrics.adoc:38]: Same comment here regarding "comes from"

File: modules/viewing-drift-metrics.adoc [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+
+[role='_abstract']
+After you create data drift monitoring metrics, you can use the {openshift-platform} web console to view and update the metrics that you configured.
```
Comment (chtyler) [inline modules/viewing-drift-metrics.adoc:7]: Are my eyes playing tricks on me, or is that a double-space after "After"?

**Review by chtyler — CHANGES_REQUESTED** [review]
A few more minor changes requested.
